File: Questions 151 - 160/152_maximumProductSubarray.py
import math


def maxProduct(nums):

    res = float(-math.inf)
    curr_max = 1
    curr_min = 1

    for num in nums:
        # update the products
        temp_max = curr_max * num
        temp_min = curr_min * num

        curr_max = max(temp_max, temp_min, num)
        curr_min = min(temp_max, temp_min, num)

        res = max(curr_max, curr_min, res)

        # reset if 0 encountered
        if(curr_max == 0):
            curr_max = 1
        if(curr_min == 0):
            curr_min = 1

    return res


if __name__ == "__main__":
    nums = [2, 3, -2, 4]

    print(f"Maximum product subarray is: {maxProduct(nums)}")

File: Questions 101 - 200/Questions 151 - 160/152_maximumProductSubarray.cpp
#include<iostream>
#include<vector>
using namespace std;

int maxProduct(vector<int>& nums) {
    int res = nums[0];
    int n = nums.size();
    int l = 1, r = 1;
    
    for(int i=0; i<n; i++){
        l = l * nums[i];
        r = r * nums[n-i-1];
        
        res = max(res, max(l,r));
        if(l==0) l = 1;
        if(r==0) r = 1;
    }
    
    return res;
}

int max(int a, int b){
    return (a>b) ? a : b;
}

int main(){
    int n;
    cout<<"Enter number of elements: ";
    cin>>n;

    vector<int> arr;
    int ele;
    cout<<"Enter array elements: ";
    for(int i=0; i<n; i++){
        cin>>ele;
        arr.push_back(ele);
    }

    cout<<"Maximum product subarray: ";
    cout<<maxProduct(arr);
    return 0;
}